packages/flutter/lib/src/widgets/widget_span.dart

```diff
@@ -0,0 +1,262 @@
+///         )
+///       ),
+///     )
+///   ),
+///   TextSpan(text: 'the best!'),
+/// ]
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [TextSpan], a node that represents text in a [TextSpan] tree.
+///  * [Text], a widget for showing uniformly-styled text.
+///  * [RichText], a widget for finer control of text rendering.
+///  * [TextPainter], a class for painting [TextSpan] objects on a [Canvas].
+@immutable
+class WidgetSpan extends PlaceholderSpan {
+  /// Creates a [WidgetSpan] with the given values.
+  ///
+  /// The [widget] property should be non-null. [WidgetSpan] cannot contain any
+  /// [TextSpan] or [WidgetSpan] children.
+  ///
+  /// A [TextStyle] may be provided with the [style] property, but only the
+  /// decoration, foreground, background, and spacing options will be used.
+  const WidgetSpan({
+    Widget widget,
+    InlineWidgetAlignment alignment = InlineWidgetAlignment.bottom,
+    TextBaseline baseline,
+    TextStyle style,
+    List<InlineSpan> children,
+  }) : assert(widget != null),
+       assert((alignment == InlineWidgetAlignment.aboveBaseline ||
+               alignment == InlineWidgetAlignment.belowBaseline ||
+               alignment == InlineWidgetAlignment.baseline) ? baseline != null : true),
+       widget = widget,
+       super(alignment:
+           // Convert InlineWidgetAlignment to PlaceholderAlignment in a const fashion.
+           alignment == InlineWidgetAlignment.baseline ? ui.PlaceholderAlignment.baseline :
+           alignment == InlineWidgetAlignment.aboveBaseline ? ui.PlaceholderAlignment.aboveBaseline :
+           alignment == InlineWidgetAlignment.belowBaseline ? ui.PlaceholderAlignment.belowBaseline :
+           alignment == InlineWidgetAlignment.top ? ui.PlaceholderAlignment.top :
+           alignment == InlineWidgetAlignment.bottom ? ui.PlaceholderAlignment.bottom :
+           alignment == InlineWidgetAlignment.middle ? ui.PlaceholderAlignment.middle : null,
+         baseline: baseline, style: style, children: children,
+       );
+
+  /// The widget to embed inline with text.
+  final Widget widget;
+
+  /// Adds a placeholder box to the paragraph builder if a size has been
+  /// calculated for the widget.
+  ///
+  /// Sizes are provided through [dimensions], which should contain a 1:1
+  /// in-order mapping of widget to laid out dimensions. If no such dimension
+  /// is provided, the widget will be skipped.
+  ///
+  /// Since widget sizes are calculated independently from the rest of the
+  /// paragraph, the [textScaleFactor] is ignored.
+  @override
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions }) {
+    assert(debugAssertIsValid());
+
+    final bool hasStyle = style != null;
+    if (hasStyle)
+      builder.pushStyle(style.getTextStyle(textScaleFactor: textScaleFactor));
+    if (dimensions != null && builder.placeholderCount < dimensions.length) {
```

Comment:
i think we should just assert `builder.placeholderCount < dimensions.length`, since if it's not something horrible has gone wrong